plugin/pkg/scheduler/algorithm/predicates/predicates.go

```diff
@@ -735,3 +735,188 @@ func haveSame(a1, a2 []string) bool {
 	}
 	return false
 }
+
+type PodAffinityChecker struct {
+	info      NodeInfo
+	podLister algorithm.PodLister
+}
+
+func NewPodAffinityPredicate(info NodeInfo, podLister algorithm.PodLister) algorithm.FitPredicate {
+	selector := &PodAffinityChecker{
+		info:      info,
+		podLister: podLister,
+	}
+	return selector.InterPodAffinityMatches
+}
+
+func (n *PodAffinityChecker) InterPodAffinityMatches(pod *api.Pod, nodeName string, nodeInfo *schedulercache.NodeInfo) (bool, error) {
+	node, err := n.info.GetNodeInfo(nodeName)
+	if err != nil {
+		return false, err
+	}
+	allPods, err := n.podLister.List(labels.Everything())
+	if err != nil {
+		return false, err
+	}
+	return nodeMatchPodAffinityAntiAffinity(pod, nodeInfo.Pods(), node, allPods), nil
+}
+
+// checkAnyPodThatMatchPodAffinityTerm checks if any of existing pods on the node can match the specific podAffinityTerm.
+func checkAnyPodThatMatchPodAffinityTerm(pod *api.Pod, existingPods []*api.Pod, node *api.Node, podAffinityTerm api.PodAffinityTerm) (bool, error) {
+	labelSelector, err := unversioned.LabelSelectorAsSelector(podAffinityTerm.LabelSelector)
+	if err != nil {
+		return false, err
+	}
+
+	names := priorityutil.GetNamespacesFromPodAffinityTerm(pod, podAffinityTerm)
+	filteredPods := priorityutil.FilterPodsByNameSpaces(names, existingPods)
+	// return true if any pod matches the podAffinityTerm
+	for _, filteredPod := range filteredPods {
+		if labelSelector.Matches(labels.Set(filteredPod.Labels)) {
+			return priorityutil.IfNodeHasTopologyKey(node, podAffinityTerm.TopologyKey), nil
+		}
+	}
+	return false, nil
+}
+
+// Checks whether the given node has pods which satisfy all the required pod affinity scheduling rules.
+// If node has pods which satisfy all the required pod affinity scheduling rules then return true.
+func checkNodeMatchesHardPodAffinity(pod *api.Pod, existingPods []*api.Pod, node *api.Node, allPods []*api.Pod, podAffinity *api.PodAffinity) bool {
```

Comment:
The function checkAnyPodThatMatchPodAffinityTerm() only looks at pods on that node. But it needs to look at all pods. For example, say you have a cluster with two nodes

node1
- node's value of "zone" label is "foo"
- node is running a pod with some set of labels L

node 2
- node's value of "zone" label is "foo"
- node is not running any pods

Now someone wants to schedule a new pod with PodAffinity with
```
PodAffinityTerm {
   LabelSelector: L
   TopologyKey: "zone"
}
```

This pod should be able to run on *both* nodes. I believe your current code will only allow it to run on node1. (There are some examples like this in the API section of the design doc, right after the API.)

I think the fix for this may be simple:

1) instead of existingPods argument to checkAnyPodThatMatchPodAffinityTerm(), use allPods as the argument

2) priorityutil.IfNodeHasTopologyKey() should take an extra argument, which is the node where filteredPod is running. So it would look something like this:
```
  func IfNodeHasTopologyKey(filteredPodNode *api.Node, node *api.Node, topologyKey string) bool {
    if len(topologyKey) == 0 { return true }
    if node.Labels != nil && filteredPodNode.Labels != nil && node.Labels[topologyKey] == filteredPodNode.Labels[topologyKey] { return true }
    return false
  }
```
Obviously there are significant performance implications for iterating over all nodes, but we can fix that later. :-)